frontend/src/app/offers/create-offer/create-offer.component.ts

```diff
@@ -0,0 +1,196 @@
+import { AppOffer, ApiOffer, BaseOffer } from 'app/homepage-offer/offers.model';
+import { AuthService } from 'app/auth.service';
+import { User } from 'app/user';
+import { Observable } from 'rxjs/Observable';
+import { OffersService } from 'app/homepage-offer/offers.service';
+import { ActivatedRoute, Params, Router } from '@angular/router';
+import { environment } from 'environments/environment';
+import { Subscription } from 'rxjs/Subscription';
+import { FileReaderEvent, FileReaderEventTarget } from '../../models';
+import { NgForm, FormGroup } from '@angular/forms';
+import { FormBuilder } from '@angular/forms';
+import { Validators } from '@angular/forms';
+import { Image } from 'app/homepage-offer/image.model';
+import { Response } from '@angular/http';
+import { HttpClient } from '@angular/common/http';
+@Component({
+  selector: 'volontulo-create-offer',
+  templateUrl: './create-offer.component.html',
+  styleUrls: ['./create-offer.component.scss']
+})
+export class CreateOfferComponent implements OnInit, OnDestroy {
+  public djangoRoot = environment.djangoRoot;
+  public hasOrganization = false;
+  public inEditMode = false;
+  public isAdmin = false;
+  public isFileToBig = false;
+  public offer: AppOffer = new AppOffer;
+  public reader = new FileReader();
+  public user: User;
+  public userSubscription: Subscription;
+  public form: FormGroup;
+  public error;
+
+  constructor(
+    private authService: AuthService,
+    private offersService: OffersService,
+    private activatedRoute: ActivatedRoute,
+    private fb: FormBuilder,
+    private http: HttpClient,
+    private router: Router,
+  ) {}
+
+  ngOnInit() {
+    this.form = this.fb.group({
+      title: ['', Validators.required],
+      location: ['', Validators.required],
+      organization: ['', Validators.required],
+      startedAt: ['', Validators.required],
+      actionOngoing: [false],
+      finishedAt: ['', Validators.required],
+      constantCoop: [false],
+      recruitmentStartDate: [null],
+      recruitmentEndDate: [null],
+      volunteersLimit: [],
+      reserveRecruitmentStartDate: [null],
+      reserveRecruitmentEndDate: [null],
+      reserveVolunteersLimit: [],
+      description: ['', Validators.required],
+      timeCommitment: ['', Validators.required],
+      benefits: ['', Validators.required],
+      requirements: [''],
+      image: [],
+    })
+
+      this.userSubscription = this.authService.user$
+      .subscribe(user => {
+        if (user) {
+          this.user = user;
+          this.isAdmin = user['isAdministrator'];
+          this.hasOrganization = user['organizations'].length > 0;
+        }
+     });
+
+    this.activatedRoute.params
+    .map(params => params.offerId)
+    .filter(offerId => offerId !== undefined)
+    .switchMap(offerId => this.offersService.getOffer(offerId))
+    .do(offer => this.toDataUrl(offer.image, (base64) => this.offer.image = {
```

Comment:
remove the callback function from here and move it inside toDataUrl(), refactor toDataUrl to use only one parameter - offer image. Then this call would look like this.toDataUrl(offer.image)